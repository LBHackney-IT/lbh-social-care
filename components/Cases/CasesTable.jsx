import PropTypes from 'prop-types';
import { formatDate } from 'utils/date';

import CaseLink from './CaseLink';

const CasesEntry = ({
  recordId,
  formName,
  caseFormUrl,
  officerEmail,
  dateOfEvent,
  caseFormData,
  caseFormTimestamp,
}) => (
  <tr className="govuk-table__row">
    <td className="govuk-table__cell govuk--timestamp">
      {formatDate(dateOfEvent || caseFormTimestamp)}{' '}
    </td>
    <td className="govuk-table__cell">{formName}</td>
    <td className="govuk-table__cell">
      {officerEmail && `- created by ${officerEmail}`}
    </td>
    <td className="govuk-table__cell govuk-button--secondary'">
      <CaseLink
        recordId={recordId}
        externalUrl={caseFormUrl}
        caseFormData={caseFormData}
      />
    </td>
  </tr>
);

CasesEntry.propTypes = {
  recordId: PropTypes.string.isRequired,
  formName: PropTypes.string.isRequired,
  caseFormUrl: PropTypes.string.isRequired,
  officerEmail: PropTypes.string.isRequired,
  dateOfEvent: PropTypes.string,
  caseFormTimestamp: PropTypes.string,
  caseFormData: PropTypes.object.isRequired,
};

const CasesTable = ({ records }) => (
  <table className="govuk-table">
    <tbody className="govuk-table__body">
      {records.map((result) => (
        <CasesEntry key={result.recordId} {...result} />
      ))}
    </tbody>
  </table>
);

CasesTable.propTypes = {
  records: PropTypes.arrayOf(PropTypes.shape(CasesEntry.propTypes)).isRequired,
};

export default CasesTable;
