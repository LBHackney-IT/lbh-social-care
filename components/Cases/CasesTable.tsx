import { Case } from 'types';
import { formatDate, isDateValid } from 'utils/date';

import CaseLink from './CaseLink';

const CaseNotePersonId = ({ personId }: Case) => (
  <td key="personId" className="govuk-table__cell">
    {personId}
  </td>
);

const CaseNoteDate = ({ dateOfEvent, caseFormTimestamp }: Case) => (
  <td
    key="date"
    className="govuk-table__cell govuk--timestamp"
    style={{ width: '120px' }}
  >
    {(dateOfEvent && formatDate(dateOfEvent)) ||
      (caseFormTimestamp && formatDate(caseFormTimestamp))}
  </td>
);

const CaseNoteTitle = ({ formName, caseFormData }: Case) => (
  <td key="title" className="govuk-table__cell">
    {
      <>
        {['ASC_case_note', 'CFS_case_note'].includes(
          caseFormData.form_name_overall
        ) && 'Case Note - '}
        {formName}
        {caseFormData.case_note_title && (
          <div>{caseFormData.case_note_title}</div>
        )}
      </>
    }
  </td>
);

const CaseNoteOfficer = ({ officerEmail }: Case) => (
  <td key="officer" className="govuk-table__cell">
    {officerEmail}
  </td>
);

const CaseNoteAction = ({ recordId, caseFormUrl, caseFormData }: Case) => (
  <td
    key="action"
    className="govuk-table__cell govuk-table__cell--numeric"
    style={{ width: '50px', textAlign: 'center' }}
  >
    <CaseLink
      recordId={recordId}
      externalUrl={caseFormUrl}
      caseFormData={caseFormData}
    />
  </td>
);

const CaseNoteName = ({ firstName, lastName }: Case) => (
  <td key="name" className="govuk-table__cell">
    {firstName} {lastName}
  </td>
);

const CaseNoteBirthday = ({ dateOfBirth }: Case) => (
  <td key="birthday" className="govuk-table__cell">
    {dateOfBirth && isDateValid(dateOfBirth) && dateOfBirth}
  </td>
);

const tableEntities = {
  person_id: { text: 'Person id', component: CaseNotePersonId },
  first_name: { text: 'Client Name', component: CaseNoteName },
  date_of_birth: { text: 'Date of birth', component: CaseNoteBirthday },
  date_of_event: { text: 'Date created', component: CaseNoteDate },
  formName: { text: 'Record type', component: CaseNoteTitle },
  officer_email: { text: 'Created by', component: CaseNoteOfficer },
  action: { text: 'Action', component: CaseNoteAction },
};

export type CaseTableColumns = keyof typeof tableEntities;

interface Props {
  records: Case[];
  columns: CaseTableColumns[];
  sort?: {
    sort_by?: string;
    order_by?: string;
  };
  onSort?: (id: string) => void;
}

const CasesTable = ({
  records,
  columns,
  sort = {},
  onSort,
}: Props): React.ReactElement => (
  <table className="govuk-table">
    <thead className="govuk-table__head">
      <tr className="govuk-table__row">
        {columns.map((column) => (
          <th
            key={column}
            scope="col"
            className="govuk-table__header"
            role={onSort && 'button'}
            onClick={() => onSort && onSort(column)}
          >
            {tableEntities[column].text}{' '}
            {column === sort.sort_by && (
              <>{sort.order_by === 'desc' ? 'ðŸ”½' : 'ðŸ”¼'}</>
            )}
          </th>
        ))}
      </tr>
    </thead>
    <tbody className="govuk-table__body">
      {records.map((record) => (
        <tr className="govuk-table__row" key={record.recordId}>
          {columns.map((column) => tableEntities[column].component(record))}
        </tr>
      ))}
    </tbody>
  </table>
);

export default CasesTable;
