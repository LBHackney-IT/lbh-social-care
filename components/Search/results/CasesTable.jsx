import PropTypes from 'prop-types';

import { formatDate, isDateValid } from 'utils/date';

const onClick = (url) => window.open(url, '_blank');

const CasesEntry = ({
  firstName,
  lastName,
  caseFormUrl,
  dateOfBirth,
  officerEmail,
  caseFormTimestamp,
}) => (
  <tr className="govuk-table__row">
    <td className="govuk-table__cell">
      {firstName} {lastName}
    </td>
    <td className="govuk-table__cell">
      {isDateValid(dateOfBirth) && dateOfBirth}
    </td>
    <td className="govuk-table__cell">{officerEmail}</td>
    <td className="govuk-table__cell">{formatDate(caseFormTimestamp)}</td>
    <td className="govuk-table__cell govuk-button--secondary'">
      <a
        href="#"
        className="govuk-link govuk-custom-text-color"
        onClick={() => onClick(caseFormUrl)}
      >
        View
      </a>
    </td>
  </tr>
);

const tableHeader = [
  { id: 'client_name', text: 'Client Name' },
  { id: 'date_of_birth', text: 'Date of birth' },
  { id: 'uploaded_by', text: 'Uploaded by' },
  { id: 'last_upload', text: 'Last upload' },
];

const CasesTable = ({ records, sort = {}, onSort }) => {
  return (
    <table className="govuk-table">
      <thead className="govuk-table__head">
        <tr className="govuk-table__row">
          {tableHeader.map(({ id, text }) => (
            <th
              key={id}
              scope="col"
              className="govuk-table__header"
              role="button"
              onClick={() => onSort(id)}
            >
              {text}{' '}
              {id === sort.sort_by && (
                <>{sort.order_by === 'desc' ? 'ðŸ”½' : 'ðŸ”¼'}</>
              )}
            </th>
          ))}
          <th scope="col" className="govuk-table__header"></th>
        </tr>
      </thead>
      <tbody className="govuk-table__body">
        {records.map((result) => (
          <CasesEntry key={result.personId} {...result} />
        ))}
      </tbody>
    </table>
  );
};

CasesTable.propTypes = {
  sort: PropTypes.shape({
    sort_by: PropTypes.string.isRequired,
    order_by: PropTypes.string.isRequired,
  }),
  onSort: PropTypes.func.isRequired,
  records: PropTypes.arrayOf(
    PropTypes.shape({
      personId: PropTypes.number.isRequired,
      formName: PropTypes.string.isRequired,
      caseFormUrl: PropTypes.string.isRequired,
      officerEmail: PropTypes.string.isRequired,
      caseFormTimestamp: PropTypes.string.isRequired,
    })
  ).isRequired,
};

export default CasesTable;
