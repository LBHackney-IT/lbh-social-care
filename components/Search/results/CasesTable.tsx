import { formatDate, isDateValid } from 'utils/date';

import CaseLink from 'components/Cases/CaseLink';
import { Case } from 'types';

const CasesEntry = ({
  recordId,
  firstName,
  lastName,
  caseFormUrl,
  dateOfBirth,
  officerEmail,
  dateOfEvent,
  caseFormTimestamp,
  caseFormData,
}: Case): React.ReactElement => (
  <tr className="govuk-table__row">
    <td className="govuk-table__cell">
      {firstName} {lastName}
    </td>
    <td className="govuk-table__cell">
      {dateOfBirth && isDateValid(dateOfBirth) && dateOfBirth}
    </td>
    <td className="govuk-table__cell">{officerEmail}</td>
    <td className="govuk-table__cell">
      {(dateOfEvent && formatDate(dateOfEvent)) ||
        (caseFormTimestamp && formatDate(caseFormTimestamp))}
    </td>
    <td className="govuk-table__cell govuk-button--secondary'">
      <CaseLink
        recordId={recordId}
        externalUrl={caseFormUrl}
        caseFormData={caseFormData}
      />
    </td>
  </tr>
);

const tableHeader = [
  { id: 'first_name', text: 'Client Name' },
  { id: 'date_of_birth', text: 'Date of birth' },
  { id: 'officer_email', text: 'Uploaded by' },
  { id: 'date_of_event', text: 'Last upload' },
];

interface Props {
  sort?: {
    sort_by?: string;
    order_by?: string;
  };
  onSort?: (id: string) => void;
  records: Case[];
}

const CasesTable = ({
  records,
  sort = {},
  onSort,
}: Props): React.ReactElement => {
  return (
    <table className="govuk-table">
      <thead className="govuk-table__head">
        <tr className="govuk-table__row">
          {tableHeader.map(({ id, text }) => (
            <th
              key={id}
              scope="col"
              className="govuk-table__header"
              role={onSort && 'button'}
              onClick={() => onSort && onSort(id)}
            >
              {text}{' '}
              {id === sort.sort_by && (
                <>{sort.order_by === 'desc' ? 'ðŸ”½' : 'ðŸ”¼'}</>
              )}
            </th>
          ))}
          <th scope="col" className="govuk-table__header"></th>
        </tr>
      </thead>
      <tbody className="govuk-table__body">
        {records.map((result) => (
          <CasesEntry key={result.recordId} {...result} />
        ))}
      </tbody>
    </table>
  );
};

export default CasesTable;
