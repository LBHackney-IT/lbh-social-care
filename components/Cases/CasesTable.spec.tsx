import { render } from '@testing-library/react';

import CasesTable from './CasesTable';
import {
  mockedNote,
  mockedCaseNote,
  mockedAllocationNote,
} from 'factories/cases';

describe('CasesTable component', () => {
  const props = {
    records: [mockedNote, mockedAllocationNote, mockedCaseNote],
  };
  it('should render properly', () => {
    const { asFragment } = render(<CasesTable {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
