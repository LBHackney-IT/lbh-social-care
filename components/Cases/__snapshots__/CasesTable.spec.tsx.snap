// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CasesTable component should render properly 1`] = `
<DocumentFragment>
  <table
    class="govuk-table"
  >
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Date created 
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Record type 
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Created by 
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Action 
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell govuk--timestamp"
          style="width: 120px;"
        >
          25 Oct 2020
        </td>
        <td
          class="govuk-table__cell"
        >
          foorm
        </td>
        <td
          class="govuk-table__cell"
        >
          Fname.Lname@hackney.gov.uk
        </td>
        <td
          class="govuk-table__cell govuk-button--secondary'"
        >
          <a
            class="govuk-link"
            href="https://foo.bar"
            rel="noreferrer noopener"
            target="_blank"
          >
            View
          </a>
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell govuk--timestamp"
          style="width: 120px;"
        >
          25 Oct 2020
        </td>
        <td
          class="govuk-table__cell"
        >
          Worker allocated
        </td>
        <td
          class="govuk-table__cell"
        >
          Fname.Lname@hackney.gov.uk
        </td>
        <td
          class="govuk-table__cell govuk-button--secondary'"
        >
          <a
            href="/people/123/allocations/321?recordId=2"
          >
            View
          </a>
        </td>
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell govuk--timestamp"
          style="width: 120px;"
        >
          25 Oct 2020
        </td>
        <td
          class="govuk-table__cell"
        >
          Case Note - foorm
          <div>
            i am a case title
          </div>
        </td>
        <td
          class="govuk-table__cell"
        >
          Fname.Lname@hackney.gov.uk
        </td>
        <td
          class="govuk-table__cell govuk-button--secondary'"
        >
          <a
            href="/people/123/records/4"
          >
            View
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
